### PR TITLE
[TOSA] Replace Linear lowering of using Matmul with Conv2d

### PR DESCRIPTION
**Pull request by Jerry-Ge**
- The existing Vela compiler doesn't support Matmul and TOSA.Fully_Connected will be deprecated
- Also add support for linear layers with rank>2 + tests

### CODE REVIEW

File: backends/arm/test/test_models.py [hunk 1]
```diff
@@ -128,13 +128,29 @@ def forward(self, x, y):
     @register_test
     class simple_linear(torch.nn.Module):
         inputs = {
-            TosaProfile.BI: (torch.ones(100, 20),),
-            TosaProfile.MI: (torch.ones(100, 20),),
+            TosaProfile.BI: (torch.ones(1, 2),),
```
Comment (digantdesai) [inline backends/arm/test/test_models.py:131]: Use rand() instead of ones()?

File: backends/arm/test/test_models.py [hunk 2]
```diff
@@ -128,13 +128,29 @@ def forward(self, x, y):
...
+            self.fc = torch.nn.Linear(2, 3)
+
+        def forward(self, x):
+            x = self.fc(x)
+            return x
+
+    @register_test
+    class simple_linear_rank4(torch.nn.Module):
+        inputs = {
+            TosaProfile.BI: (torch.ones(5, 10, 100, 20),),
```
Comment (digantdesai) [inline backends/arm/test/test_models.py:147]: 100 to something smaller?

**Review by digantdesai — COMMENTED** [review]
LGTM at a high level, let's close the comments?

File: backends/arm/tosa_utils.py [hunk 3]
```diff
@@ -132,3 +135,33 @@ def buildReshape(tosa_fb, input_name, new_shape, output_name):
     attr = ts.TosaSerializerAttribute()
     attr.ReshapeAttribute(new_shape)
     tosa_fb.addOperator(TosaOp.Op().RESHAPE, [input_name], [output_name], attr)
+
+
+def isPermuteNodeBeforeAddMM(node):
```
Comment (robell) [inline backends/arm/tosa_utils.py:140]: This should follow the pep008 conventions and be lower case with underscores. same for isBiasNodeForAdMM below. (I know we're not exclusively correct in the existing code but we can move towards it).

File: backends/arm/operators/op_addmm.py [hunk 4]
```diff
@@ -37,69 +39,84 @@ def define_node(
...
+        conv2d_res = tosa_graph.addIntermediate(
+            conv2d_output_shape,
+            ts.DType.INT32 if is_quant_node else output.dtype,
+        )
         tosa_graph.addOperator(
-            TosaOp.Op().ADD,
-            [bias_reshape_res.name, mm_res.name],
-            [add_res.name],
-            None,
+            TosaOp.Op().CONV2D,
```
[the remark this answers is not in the document]
Comment (robell) [reply backends/arm/operators/op_addmm.py:77]: U55 doesn't support matmul from two activation channels (it has operations which are activation, weights as input). This should also be faster generally as weight inputs are typically optimised.

File: backends/arm/operators/op_conv2d.py [hunk 5]
```diff
@@ -52,7 +52,7 @@ def define_node(
         pad_attr = [val for val in pad.special for _ in (0, 1)]
         stride_attr = stride.special
         dilation_attr = dilation.special
-        attr.ConvAttribute(pad_attr, stride_attr, dilation_attr, 0, 0)
+        attr.ConvAttribute(pad_attr, stride_attr, dilation_attr, 0, 0, False)
```
[the remark this answers is not in the document]
Comment (robell) [reply backends/arm/operators/op_conv2d.py:55]: I modified serialization_lib upstream to have basic packaging so we can pin a version and move this in sync while it's dev quality. We're expecting from 1.0 (we're 0.9 now) that there will be some compatibility guarantees for the TOSA spec and so this should hold for the serialization_library (which various other codebases rely on).

[the remark this answers is not in the document]
Comment (digantdesai) [reply backends/arm/operators/op_conv2d.py:55]: Explicit versions are good, thanks Robert. 

@Jerry-Ge  feel free to resolve this after adding arg names.

File: backends/arm/operators/op_addmm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (digantdesai) [reply backends/arm/operators/op_addmm.py:77]: Ok. IIUC, should we lower something like torch.[b]mm to this as well?

File: backends/arm/operators/op_addmm.py [hunk 6]
```diff
@@ -37,69 +39,84 @@ def define_node(
     ) -> None:
         bias, input, weight = inputs
 
-        output_dtype = ts.DType.INT8 if is_quant_node else output.dtype
+        N = input.shape[0]
+        IC = input.shape[1]
```
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply backends/arm/operators/op_addmm.py]: done

File: backends/arm/operators/op_addmm.py [hunk 7]
```diff
@@ -37,69 +39,84 @@ def define_node(
...
+                input_reshaped.name,
+                weight_reshaped.name,
+                bias.name,
+            ],
+            [conv2d_res.name],
+            attr,
         )
 
+        result_shape = (N, OC)
+        if is_quant_node:
```
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply backends/arm/operators/op_addmm.py]: done.

File: backends/arm/test/test_models.py (hunk 1 above)
Comment (Jerry-Ge) [reply backends/arm/test/test_models.py]: done.

File: backends/arm/test/test_models.py (hunk 2 above)
Comment (Jerry-Ge) [reply backends/arm/test/test_models.py]: done.

File: backends/arm/tosa_utils.py (hunk 3 above)
Comment (Jerry-Ge) [reply backends/arm/tosa_utils.py]: agree. refactored those two (isPermuteNodeBeforeAddMM and isBiasNodeForAdMM). will do another full path to refactor all the relevant code.

File: backends/arm/operators/op_addmm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply backends/arm/operators/op_addmm.py:87]: > Ok. IIUC, should we lower something like torch.[b]mm to this as well?

i think so.

File: backends/arm/tosa_utils.py (hunk 3 above)
Comment (digantdesai) [reply backends/arm/tosa_utils.py]: Feel free to create issues and assign it to yourself :p

**Review by digantdesai — APPROVED** [review]
LGTM, let me know if you are ready I can merge it. Thanks @Jerry-Ge

**Comment by Jerry-Ge** [comment]
> LGTM, let me know if you are ready I can merge it. Thanks @Jerry-Ge

thanks Digant! it's ready and let's merge it.

File: backends/arm/tosa_utils.py (hunk 3 above)
Comment (Jerry-Ge) [reply backends/arm/tosa_utils.py]: yep, sounds good.

**Pull request closed without merging**